core/src/test/scala/unit/kafka/server/AddPartitionsToTxnRequestServerTest.scala

```diff
@@ -55,22 +65,149 @@ class AddPartitionsToTxnRequestServerTest extends BaseRequestTest {
+          .setPartitions(Collections.singletonList(createdTopicPartition.partition())))
+        topics.add(new AddPartitionsToTxnTopic()
+          .setName(nonExistentTopic.topic())
+          .setPartitions(Collections.singletonList(nonExistentTopic.partition())))
+
+        val transactions = new AddPartitionsToTxnTransactionCollection()
+        transactions.add(new AddPartitionsToTxnTransaction()
+          .setTransactionalId(transactionalId)
+          .setProducerId(producerId)
+          .setProducerEpoch(producerEpoch)
+          .setVerifyOnly(false)
+          .setTopics(topics))
+        AddPartitionsToTxnRequest.Builder.forBroker(transactions).build()
+      }
 
     val leaderId = brokers.head.config.brokerId
     val response = connectAndReceive[AddPartitionsToTxnResponse](request, brokerSocketServer(leaderId))
+    
+    val errors = 
+      if (version < 4) 
+        response.errors.get(AddPartitionsToTxnResponse.V3_AND_BELOW_TXN_ID) 
+      else 
+        response.errorsForTransaction(response.getTransactionTopicResults(transactionalId))
+    
+    assertEquals(2, errors.size)
+
+    assertTrue(errors.containsKey(createdTopicPartition))
+    assertEquals(Errors.OPERATION_NOT_ATTEMPTED, errors.get(createdTopicPartition))
+
+    assertTrue(errors.containsKey(nonExistentTopic))
+    assertEquals(Errors.UNKNOWN_TOPIC_OR_PARTITION, errors.get(nonExistentTopic))
+  }
+  
+  @Test
+  def testOneSuccessOneErrorInBatchedRequest(): Unit = {
+    val tp0 = new TopicPartition(topic1, 0)
+    val transactionalId1 = "foobar"
+    val transactionalId2 = "barfoo" // "barfoo" maps to the same transaction coordinator
+    val producerId2 = 1000L
+    val producerEpoch2: Short = 0
+    
+    val txn2Topics = new AddPartitionsToTxnTopicCollection()
+    txn2Topics.add(new AddPartitionsToTxnTopic()
+      .setName(tp0.topic())
+      .setPartitions(Collections.singletonList(tp0.partition())))
+
+    val (coordinatorId, txn1) = setUpTransactions(transactionalId1, false, Set(tp0))
+
+    val transactions = new AddPartitionsToTxnTransactionCollection()
+    transactions.add(txn1)
+    transactions.add(new AddPartitionsToTxnTransaction()
+      .setTransactionalId(transactionalId2)
+      .setProducerId(producerId2)
+      .setProducerEpoch(producerEpoch2)
+      .setVerifyOnly(false)
+      .setTopics(txn2Topics))
+
+    val request = AddPartitionsToTxnRequest.Builder.forBroker(transactions).build()
+
+    val response = connectAndReceive[AddPartitionsToTxnResponse](request, brokerSocketServer(coordinatorId))
+
+    val errors = response.errors()
+
+    assertTrue(errors.containsKey(transactionalId1))
+    assertTrue(errors.get(transactionalId1).containsKey(tp0))
+    assertEquals(Errors.NONE, errors.get(transactionalId1).get(tp0))
+
+    assertTrue(errors.containsKey(transactionalId2))
+    assertTrue(errors.get(transactionalId1).containsKey(tp0))
+    assertEquals(Errors.INVALID_PRODUCER_ID_MAPPING, errors.get(transactionalId2).get(tp0))
```

Comment:
nit: How about using `assertEquals(Map(... define the expect map ..., errors)`? This is simpler and has the benefit of ensuring that we only have what we expect in `errors`.

Note that assertion as L148 is not correct. `transactionalId2` should be used.